### DISCUSSION TITLE
Training ControlNet on Stable Diffusion Inpaint

### DISCUSSION THREAD

**Discussion by engrmusawarali** [huggingface/diffusers#6684] [General]
How do we create weights for the controlnet inpaint which uses stable diffusion inpaint as backbone @patrickvonplaten @sayakpaul @Dhruv Nair

**Comment by patrickvonplaten** [comment]
Hey @engrmusawarali,

Do you mean how to fine-tune a stable diffusion base checkpoint for controlnet inpainting?

**Reply by engrmusawarali** [reply]
@patrickvonplaten I have implemented the training code for training controlnet on stable diffusion inpaint, but I am not sure what is the best way to copy the weights from stable diffusion unet to controlnet unet. That was my query. Since the sd  inpaint has 9 channels, we need to copy first 4 channel weights, but i am wondering about what is the best way to copy the weghts, I think my query is not pretty detailed.

**Reply by sayakpaul** [reply]
I think you could follow what InstructPix2Pix did. I assume you are adding some more channels to the input conv layers in the ControlNet module to accommodate for the additional channels. So, you copy as much as possible and for the extra channels you initialize like so:

https://github.com/huggingface/diffusers/blob/5b1b80a5b6633126994636ea289393b9e931834a/examples/instruct_pix2pix/train_instruct_pix2pix.py#L464

**Reply by engrmusawarali** [reply]
@sayakpaul thanks for the reply. I dont want to add more channels in controlnet module, I only want to copy first 4-channels of sd inpaint to controlnet.


I did this to copy weights 


 unetnew = copy.deepcopy(unet)
 unetnew.register_to_config(in_channels=4)
        with torch.no_grad():
            new_conv_in = nn.Conv2d(
                4, unet.conv_in.out_channels, unet.conv_in.kernel_size, unet.conv_in.stride, unet.conv_in.padding
            )
            new_conv_in.weight.copy_(unet.conv_in.weight[:, :4, :, :])
            unetnew.conv_in = new_conv_in
            # Create a ControlNetModel from the modified UNet model
            controlnet = ControlNetModel.from_unet(unetnew)

**Reply by sayakpaul** [reply]
Seems fine. What's the problem with that?

**Reply by engrmusawarali** [reply]
@sayakpaul  i prepared this using suggestion from instruct pix2pix, I wanted to confirm If this is correct way, It would be good if this option is integrated in the pipelines. Thanks a lot for hint

**Reply by sayakpaul** [reply]
If we have good checkpoints and convincing results, then I don't see any reason not to.

**Reply by engrmusawarali** [reply]
@sayakpaul, it is totally dependent on the problem we are solving, my experience is really good, with controlnet inpaint with stable diffusion inpaint, before this I was initializing cntrolnet unet with stable diffusion base unet, But i wanted to follow the exact protocols proposed by controlnets.

**Reply by shoaibmustafakhan** [reply]
I'm working a research project myself and I would love to know if you've developed a finetuning script for the model?
